ICSharpCode.Decompiler/CSharp/Syntax/TypeSystemAstBuilder.cs

```diff
@@ -915,9 +917,276 @@ Expression ConvertEnumValue(IType type, long val)
 			}
 			return new CastExpression(ConvertType(type), new PrimitiveExpression(CSharpPrimitiveCast.Cast(enumBaseTypeCode, val, false)));
 		}
-		
+
+		static bool IsValidFraction(long num, long den)
+		{
+			if (!(den > 0 && num != 0))
+				return false;
+
+			if (den == 1 || Math.Abs(num) == 1)
+				return true;
+			return Math.Abs(num) < den && new int[] { 2, 3, 5 }.Any(x => den % x == 0);
+		}
+
+		const int MAX_DENOMINATOR = 1000;
+
+		const double DOUBLE_THRESHOLD = 1e-10;
+		const float FLOAT_THRESHOLD = 1e-10f;
```

Comment:
What's this?
Float literals are exact, we certainly shouldn't use a constant if the value does not match exactly.

Reply:
> Float literals are exact

That's what I thought, until I discovered this:

```csharp
        const double Doubles01 = 3.1415926535897931;
        const double Doubles02 = 3.141592653589793;
        const double Doubles03 = 3.14159265358979;
        const double Doubles04 = 3.1415926535897;
        const double Doubles05 = 3.141592653589;
        const double Doubles06 = 3.14159265358;
        const double Doubles07 = 3.1415926535;
        const double Doubles08 = 3.141592653;
        const double Doubles09 = 3.14159265;
        const double Doubles10 = 3.1415926;
        const double Doubles11 = 3.141592;
        const double Doubles12 = 3.14159;
        const double Doubles13 = 3.1415;
        const double Doubles14 = 3.141;
        const double Doubles15 = 3.14;
        const double Doubles16 = 3.1;
        const double Doubles17 = 3;

        const float Floats1 = 3.14159274f;
        const float Floats2 = 3.1415927f;
        const float Floats3 = 3.141592f;
        const float Floats4 = 3.14159f;
        const float Floats5 = 3.1415f;
        const float Floats6 = 3.141f;
        const float Floats7 = 3.14f;
        const float Floats8 = 3.1f;
        const float Floats9 = 3;
```

gets compiled to this:
```
	.field private static literal float64 Doubles01 = float64(3.1415926535897931) // ??
	.field private static literal float64 Doubles02 = float64(3.1415926535897931) // ??
	.field private static literal float64 Doubles03 = float64(3.14159265358979)
	.field private static literal float64 Doubles04 = float64(3.1415926535897)
	.field private static literal float64 Doubles05 = float64(3.141592653589)
	.field private static literal float64 Doubles06 = float64(3.14159265358)
	.field private static literal float64 Doubles07 = float64(3.1415926535)
	.field private static literal float64 Doubles08 = float64(3.141592653)
	.field private static literal float64 Doubles09 = float64(3.14159265)
	.field private static literal float64 Doubles10 = float64(3.1415926)
	.field private static literal float64 Doubles11 = float64(3.141592)
	.field private static literal float64 Doubles12 = float64(3.14159)
	.field private static literal float64 Doubles13 = float64(3.1415)
	.field private static literal float64 Doubles14 = float64(3.141)
	.field private static literal float64 Doubles15 = float64(3.14)
	.field private static literal float64 Doubles16 = float64(3.1)
	.field private static literal float64 Doubles17 = float64(3)
	.field private static literal float32 Floats1 = float32(3.14159274) // ??
	.field private static literal float32 Floats2 = float32(3.14159274) // ??
	.field private static literal float32 Floats3 = float32(3.141592)
	.field private static literal float32 Floats4 = float32(3.14159)
	.field private static literal float32 Floats5 = float32(3.1415)
	.field private static literal float32 Floats6 = float32(3.141)
	.field private static literal float32 Floats7 = float32(3.14)
	.field private static literal float32 Floats8 = float32(3.1)
	.field private static literal float32 Floats9 = float32(3)
```

Not sure what's causing this...

Reply:
Oh sure, there can be multiple decimal values that get mapped to the same float.
But it's always well-defined which float you get, and if the IL code has a similar but different float, we can't just a constant without changing the semantics of the program.